crates/sui-framework/packages/sui-framework/sources/coin.move

```diff
@@ -272,9 +283,44 @@ public fun create_regulated_currency_v2<T: drop>(
         coin_metadata_object: object::id(&metadata),
         deny_cap_object: object::id(&deny_cap),
     });
+    if (ctx.epoch() >= REGULATED_MARKER_START_EPOCH) {
+        dynamic_field::add(&mut metadata.id, RegulatedMarker {}, allow_global_pause);
+        assert!(is_regulated_currency(&metadata, ctx).extract());
+    };
     (treasury_cap, deny_cap, metadata)
 }
 
+/// Return Some(true) if `T` is a regulated currency, Some(false) if it is an ordinary currency.
+/// Return None if T was created at an epoch before `REGULATED_MARKER_START_EPOCH`. Before
+/// this epoch, it was not possible to distinguish between regulated and unregulated currencies by
+/// looking at the metadata.
+public fun is_regulated_currency<T>(metadata: &CoinMetadata<T>, ctx: &mut TxContext): Option<bool> {
```

Comment:
I don't think context should be mutable here. Can't think of a single scenario where we'd find it useful.

Reply:
Good point + I'm surprised there's no linter or compiler warning to yell at me. Is this because it would be too noisy in the case that folks do this for upgrade friendliness, and it's too hard to distinguish between that case and this one?